NIFI-9398 add verification to ElasticSearchClientService (with integration tests) and Elasticsearch REST API processors

ChrisSamo632 opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-9398](https://issues.apache.org/jira/browse/NIFI-9398) add verification for the Elasticsearch REST API processors and controller services.

NIFI-9398 processor verification integration-tests for nifi-elasticsearch-restapi-processors using common nifi-elasticsearch-integration-test-utils refactored from nifi-elasticsearch-client-service

Additionally:
- add `elasticsearch8` tags to Elasticsearch REST API processors
- address various compiler warnings (i.e. from IntelliJ) in associated classes
- refactor Elasticsearch (Testcontainers) integration-tests into new utils modules shared between client-service and restapi-processors modules
- use `testcontainers-bom` to set version of all Testcontainer imported dependencies throughout NiFi repo

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [x] JDK 11
  - [x] JDK 17

### Licensing

- ~[ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)~
- ~[ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files~

### Documentation

- ~[ ] Documentation formatting appears as expected in rendered files~

ChrisSamo632 commented:
Marked as `Draft` until I've completed more manual verifications via the NiFi UI.

ChrisSamo632 commented:
> Marked as `Draft` until I've completed more manual verifications via the NiFi UI.

Added integration-tests for `nifi-elasticsearch-restapi-processors` to cover some Processor Verification steps (shares Testcontainer setup with `nifi-elasticsearch-client-service` using a new common `nifi-elasticsearch-integration-test-utils` module)

MikeThomsen reviewed (changes requested):
A few questions and concerns, but otherwise looks really good. Tested it locally, and the changes did not work with the `integration-tests` profile activated until the test utils package got those updates.

ChrisSamo632 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/GetElasticsearch.java:
```diff
@@ -213,7 +273,7 @@ private void handleElasticsearchException(final ElasticsearchException ese, Flow
             if (input != null) {
                 session.penalize(input);
-                input = session.putAttribute(input, "elasticsearch.get.error", ese.getMessage());
+                session.putAttribute(input, "elasticsearch.get.error", ese.getMessage());
```
I was thinking that `input` wasn't then passed back out of the method and so there was no need to re-assign it, but actually it's used in the subsequent `session.transfer`, so agree this should be re-instated

ChrisSamo632 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-client-service/src/main/java/org/apache/nifi/elasticsearch/ElasticSearchClientServiceImpl.java:
```diff
@@ -416,6 +495,26 @@ public void refresh(final String index, final Map<String, String> requestParamet
+    @Override
+    public boolean exists(final String index, final Map<String, String> requestParameters) {
+        try {
+            final Response response = performRequest("HEAD", "/" + index, requestParameters, null);
```
I think this is a common thing throughout the methods in this class and was wondering the same myself (but then forgot) - I'll make a general change to ensure that a leading `/` is always present but that we don't end up with duplicates (e.g. `host:port//index`)

ChrisSamo632 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ElasticsearchRestProcessor.java:
```diff
@@ -117,13 +123,66 @@ default String getQuery(final FlowFile input, final ProcessContext context, fina
+    String VERIFICATION_STEP_INDEX_EXISTS = "Elasticsearch Index Exists";
+
+    @Override
+    default List<ConfigVerificationResult> verify(final ProcessContext context, final ComponentLog verificationLogger, final Map<String, String> attributes) {
```
Sounds like you might have been confused by the NiFi `verify` process as I was the first time I reviewed this sort of thing (I think there was a documentation Jira raised off the back of a Slack conversation I started, but not sure it's been done yet)

Verification in NiFi is an optional thing for users to test their processor/controller settings before enabling/starting their components to verify whether they've got things like connection details setup correctly. Here, I've set the "Index Exists Check" to `SUCCEED` whether the index exists or not, provided the connection to Elasticsearch works and the expected not/exists response is returned. So it's OK if the index doesn't exist (and I'd see that happening not just for templated indices but any index - the index could be created by the first use of `PutElasticsearchRecord`, for example)

Verification in NiFi is **not** mandatory and if a verify check fails, it doesn't stop the user from enabling/starting their components - but here, it should pass whether or not the index exists provided the connection to Elasticsearch succeeds and there are no authentication issues - the user is given the opportunity to provide attribute values via the verification window, e.g. if they're using NiFi Expression Language to specify an index name at runtime from FlowFile attributes, they can simulate that in the UI for verification

ChrisSamo632 replied:
To be fair, maybe it should be a `FAILED` status for processors that wouldn't create the index, e.g. Get or Query

ChrisSamo632 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/main/java/org/apache/nifi/processors/elasticsearch/ElasticsearchRestProcessor.java:
```diff
@@ -117,13 +123,66 @@ default String getQuery(final FlowFile input, final ProcessContext context, fina
+                                .explanation(String.format("Index [%s] exists", index));
+                        indexExists = true;
+                    } else {
+                        indexExistsResult.outcome(ConfigVerificationResult.Outcome.SUCCESSFUL)
```
Correct, I've tried to describe this further in https://github.com/apache/nifi/pull/6544#discussion_r1014601005

ChrisSamo632 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-test-utils/pom.xml:
```diff
@@ -0,0 +1,70 @@
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>elasticsearch</artifactId>
```
Default `scope` in Maven is `compile`, so don't need to be explicit (although we can if wanted for all dependencies)

The `version` is taken from the `testcontainers` in the top-level `nifi/pom.xml`'s `dependencyManagement` for all subsequent child POMs using `testcontainer` imports

MikeThomsen replied:
> The version is taken from the testcontainers in the top-level nifi/pom.xml's dependencyManagement for all subsequent child POMs using testcontainer imports

The root pom's `<dependencyManagement/>` section has an explicit scope of `test` set on testcontainers dependencies.

ChrisSamo632 replied:
Seemed to work for me, maybe I'd changed the scope and not rebuilt the main pom or something, no matter, will add the compile scope

ChrisSamo632 commented on nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-test-utils/pom.xml:
```diff
@@ -0,0 +1,70 @@
+            <artifactId>elasticsearch</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
```
As per https://github.com/apache/nifi/pull/6544/files#r1014601265

MikeThomsen commented:
@ChrisSamo632 see feedback above. When I ran your branch on the command line, I got a bunch of NoClassDef errors because testcontainers has an explicit scope of `test` set in the root pom.  `mvn dependency:tree` was showing that testcontainers was not coming through as a transitive dependency via the test utils jar.

ChrisSamo632 commented:
> @ChrisSamo632 see feedback above. When I ran your branch on the command line, I got a bunch of NoClassDef errors because testcontainers has an explicit scope of `test` set in the root pom. `mvn dependency:tree` was showing that testcontainers was not coming through as a transitive dependency via the test utils jar.

@MikeThomsen thanks for the review, I think I've addressed all your comments now (I unresolved those in the discussion that I believed needed changes to be made, that way I/you can see what needs re-reviewing).

Also updated the 7.x and 8.x image versions to the latest available (new releases from Elastic recently).
